lib/report.js

```diff
@@ -158,7 +158,99 @@ class ConciseReporter extends Reporter {
   }
 }
 
+class TapReporter extends Reporter {
+  constructor(options) {
+    super(options);
+    this.successful = 0;
+    this.failed = 0;
+    this.skipped = 0;
+    this.testNumeration = [0];
+    this.lastIndex = 0;
+    this.message = '';
+  }
+
+  error(test, error) {
+    // remove one success because 'error' may only be received after end
+    // though the test may have already failed so we need a way to check
+    // if we should do this, i.e. store map of successful tests
+    // this.successes--;
+    // test.todo ? this.skipped++ : this.failures++;
+    // TODO(lundibundi): add test.metadata.filename here too once available
+    this.listFailure(test,
+      { message: `Failure after end: ${error ? error.message : 'no message'}` },
+      ` Test failed: ${test.caption}`);
+  }
+
+  lastElement(inc) {
+    if (inc) this.testNumeration[this.lastIndex]++;
+    return this.testNumeration[this.lastIndex];
+  }
+
+  parseTestResults(test) {
+    const testPrefix = test.success ? 'ok' : 'not ok';
+    this.lastElement(1);
+    this.message = `${testPrefix} ${this.testNumeration[this.lastIndex]}`;
+    this.message += ` - ${test.caption}\n`;
+    const firstFailure = true;
+    const got = [];
+    const expected = [];
+    if (!test.success) {
+      this.listFailure(test, firstFailure, got, expected);
+      got.forEach(el => {
+        this.message += ' '.repeat(6);
+        this.message += `- ${el}\n`;
+      });
+      this.message += '    expect: \n';
+      expected.forEach(el => {
+        this.message += ' '.repeat(6);
+        this.message += `- ${el}\n`;
```

Comment:
This should be extracted into a separate function and `expected.forEach(serializeElement)`. And the same for `got`.